### PR TITLE
Lazy load images in the Brave News V2 UI

### PR DESCRIPTION
**Pull request by fallaciousreasoning**
Resolves https://github.com/brave/brave-browser/issues/26161

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_new_tab_ui/components/default/braveToday/useUnpaddedImageUrl.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+export function useLazyUnpaddedImageUrl (paddedUrl: string | undefined, options: Options) {
+  const [visible, setVisible] = React.useState(false)
+  const [elementRef, setElementRef] = React.useState<HTMLElement | null>(null)
+
+  React.useEffect(() => {
+    if (!elementRef) return
+
+    const observer = new IntersectionObserver(([intersectionInfo]) => {
```
Comment (sangwoo108) [inline components/brave_new_tab_ui/components/default/braveToday/useUnpaddedImageUrl.ts:76]: Wow, cool! I didn't know that there's such a thing.

File: components/brave_new_tab_ui/components/default/braveToday/useUnpaddedImageUrl.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+  observer.observe(elementRef)
+    return () => {
+      observer.disconnect()
+    }
+  }, [options.rootElement, elementRef])
+
+  return {
+    url: useUnpaddedImageUrl(visible ? paddedUrl : undefined, options.onLoaded, options.useCache) || cache[paddedUrl!],
+    elementRef: setElementRef
```
Comment (sangwoo108) [inline components/brave_new_tab_ui/components/default/braveToday/useUnpaddedImageUrl.ts:95]: nit: how about renaming this to setElementRef as it's callback ref?
```suggestion
    setElementRef: setElementRef
```

File: components/brave_new_tab_ui/components/default/braveToday/useUnpaddedImageUrl.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    // Storybook method
+    // @ts-expect-error
+    if (window.braveStorybookUnpadUrl) {
+      // @ts-expect-error
+      window.braveStorybookUnpadUrl(paddedUrl)
+        .then(onReceiveUnpaddedUrl)
+      return
+    }
+
+    if (!paddedUrl) return
```
Comment (sangwoo108) [inline components/brave_new_tab_ui/components/default/braveToday/useUnpaddedImageUrl.ts:39]: Can we move this up to the first line?

**Review by sangwoo108 — APPROVED** [review]
LGTM 👍

File: components/brave_new_tab_ui/components/default/braveToday/useUnpaddedImageUrl.ts (hunk 1 above)
Comment (fallaciousreasoning) [reply components/brave_new_tab_ui/components/default/braveToday/useUnpaddedImageUrl.ts:76]: Yeah, there are so many new Web APIs it's kinda hard to keep track, right? :laughing:

File: components/brave_new_tab_ui/components/default/braveToday/useUnpaddedImageUrl.ts (hunk 2 above)
Comment (fallaciousreasoning) [reply components/brave_new_tab_ui/components/default/braveToday/useUnpaddedImageUrl.ts:95]: :+1: I was uuuming and aahing on this

File: components/brave_new_tab_ui/components/default/braveToday/useUnpaddedImageUrl.ts (hunk 3 above)
Comment (fallaciousreasoning) [reply components/brave_new_tab_ui/components/default/braveToday/useUnpaddedImageUrl.ts:39]: I don't think so, unfortunately - the storybook override needs to work even if there isn't a url passed in (i.e. in the case of our mock data).

File: components/brave_new_tab_ui/components/default/braveToday/useUnpaddedImageUrl.ts (hunk 1 above)
Comment (sangwoo108) [reply components/brave_new_tab_ui/components/default/braveToday/useUnpaddedImageUrl.ts:76]: Yeah, I learned web apis about 8 yrs ago, and now I feel like I don't know most of apis haha. How do you keep track of them? Brave news?

**Review by sangwoo108 — APPROVED** [review]
Still LGTM 👍

**Comment by stephendonner** [comment]
## Verified `PASSED` using 

Brave | 1.47.47 Chromium: 107.0.5304.91 (Official Build) nightly (x86_64)
-- | --
Revision | 3d5948960d62418160796d5831a4d2d7d6c90fa8-refs/branch-heads/5304@{#1097}
OS | macOS Version 13.1 (Build 22C5033e)

### Steps:

1. installed `1.47.47`
2. launched Brave
3. clicked on `Skip welcome tour`
4. opened `brave://flags`
5. set `brave://flags/#brave-news-v2` to `Enabled`
6. clicked on `Relaunch`
7. opened a new-tab page
8. clicked on `Customize`
9. clicked on `Brave News`
10. clicked on `Turn on Brave News`

#### Confirmed the `Channels` icons lazily-loaded, filling in as they scrolled into view

**Pull request merged**